File: backend/app/api/games.py
from flask import Blueprint, jsonify, request
from app.models import User, BoardGame, Offer, db
import datetime


games_routes = Blueprint('games', __name__)


@games_routes.route('/remove', methods=['DELETE'])
def remove_game():
  try:
    game_id = request.args.get('game_id')
    remove = BoardGame.query.filter(BoardGame.id == game_id).delete()
    db.session.commit()
    return {'Message': "Successfully removed."}
  except:
    return {"Message": "Could not delete game."}

@games_routes.route('/collection')
def get_user_collection():
  user_id = request.args.get('id')
  games = BoardGame.query.filter(BoardGame.user_id == user_id).order_by(BoardGame.title)
  user_info = User.query.filter(User.id == user_id).one()
  data = [game.to_dict() for game in games]
  return {"games": data, "gameOwner": user_info.to_dict()}, 200

@games_routes.route('/borrowedgames')
def get_borrowed_games():
  user_id = request.args.get('id')
  games = BoardGame.query.filter(BoardGame.borrower_id == user_id).order_by(BoardGame.title)
  data = [game.to_dict() for game in games]
  return {"games": data}

@games_routes.route('/offer')
def get_offer_game():
  game_id = request.args.get('id')
  game = BoardGame.query.filter(BoardGame.id == game_id).one()
  return {"game": game.to_dict()}

@games_routes.route('/collection/sale')
def get_user_buy():
  user_id = request.json.get('id')
  games = BoardGame.query.filter(BoardGame.user_id == user_id).with_entities(
    BoardGame.forsale == True).all()
  data = [game.to_dict() for game in games]
  return {"games": data}, 200

@games_routes.route('/collection/trade')
def get_user_trade():
  user_id = request.json.get('id')
  games = BoardGame.query.filter(BoardGame.user_id == user_id).with_entities(
    BoardGame.fortrade == True).all()
  data = [game.to_dict() for game in games]
  return {"games": data}, 200

@games_routes.route('/collection/borrow')
def get_user_borrow():
  user_id = request.json.get('id')
  games = BoardGame.query.filter(BoardGame.user_id == user_id).with_entities(
    BoardGame.forborrow == True).all()
  data = [game.to_dict() for game in games]
  return {"games": data}, 200

@games_routes.route('/forbuy')
def get_games_for_buy():
  games = BoardGame.query.filter(BoardGame.forsale == True).all()
  data = [game.to_dict() for game in games]
  return {"games": data}, 200

@games_routes.route('/fortrade')
def get_games_for_trade():
  games = BoardGame.query.filter(BoardGame.fortrade == True).all()
  data = [game.to_dict() for game in games]
  return {"games": data}, 200

@games_routes.route('/forborrow')
def get_games_for_borrow():
  games = BoardGame.query.filter(BoardGame.forborrow == True).all()
  data = [game.to_dict() for game in games]
  return {"games": data}, 200

@games_routes.route('/bytitle')
def get_games_by_title():
  # try:
  user_id = request.args.get('id')
  search_term = request.args.get('searchTerm')
  games = BoardGame.query.filter(BoardGame.user_id == user_id, BoardGame.title.startswith(search_term)).limit(4)
  data = [game.to_dict() for game in games]
  return {"games": data}, 200
  # except:
    # return make_response({"msg": "Error. Could not send response."})

@games_routes.route('/add', methods=['POST'])
def add_game_to_collection():
  try:
    userId = request.json.get('user_id')
    gameId = request.json.get('game_id')
    username = request.json.get('username')
    game_title = request.json.get('title')
    year_published = request.json.get('year_published')
    thumb_url = request.json.get('thumb_url')
    rank = request.json.get('rank')
    game_condition = request.json.get('gameCondition')
    game_description = request.json.get('conditionDescription')
    list_price = float(request.json.get('listingPrice'))
    game_msrp = float(request.json.get('msrp'))
    checked_sale = bool(request.json.get('forsale'))
    checked_trade = bool(request.json.get('fortrade'))
    checked_borrow = bool(request.json.get('forborrow'))
    game = BoardGame(
      user_id=userId,
      game_id=gameId,
      username=username,
      title=game_title,
      year_published=year_published,
      thumb_url=thumb_url,
      msrp=game_msrp,
      sale_price=list_price,
      rank=rank,
      forsale=checked_sale,
      fortrade=checked_trade,
      forborrow=checked_borrow,
      condition=game_condition,
      condition_description=game_description,
    )
    db.session.add(game)
    db.session.commit()
    return {"Message": "Games successfully added"}, 200
  except:
    return {"Message": sqlalchemy.exc.DataError}, 500

@games_routes.route('/update', methods=['PATCH'])
def update_game():
  try:
    game_id = request.args.get('game_id')
    list_price = float(request.json.get('listingPrice'))
    game_condition = request.json.get('gameCondition')
    game_description = request.json.get('conditionDescription')
    checked_sale = bool(request.json.get('forsale'))
    checked_trade = bool(request.json.get('fortrade'))
    checked_borrow = bool(request.json.get('forborrow'))

    update = BoardGame.query.filter(BoardGame.id == game_id).first()

    update.sale_price=list_price
    update.condition=game_condition
    update.condition_description=game_description
    update.forsale=checked_sale
    update.fortrade=checked_trade
    update.forborrow=checked_borrow

    db.session.commit()
    return {'Message': 'Successfully updated game'}
  except:
    return {'Message': 'Unable to update game.'}

@games_routes.route('/toggleforsale', methods=['PATCH'])
def toggle_for_sale():
  user_id = request.json.get('user_id')
  game_tag = request.json.get('game_id')
  game = BoardGame.query.filter(BoardGame.user_id == user_id, BoardGame.game_id == game_tag).first()
  if (game.forsale == False):
    game.forsale = True
  db.session.commit()
  return {"Message": "Games successfully listed for sale."}

@games_routes.route('/changeowner', methods=["PATCH"])
def change_owner():
  try:
    new_owner_id = request.json.get('offeree_id')
    game_id = request.json.get('game_id')
    username = request.json.get('username')
    game = BoardGame.query.filter(BoardGame.id == game_id).first()

    game.user_id = new_owner_id
    game.username = username
    game.forsale = False
    game.fortrade = False
    game.forborrow = False
    game.condition_description = None
    game.sale_price = 0
    db.session.commit()
    return {"Message": "Transaction compelted."}
  except:
    return {"Message": "Transaction could not complete."}

@games_routes.route('/borrow', methods=["PATCH"])
def borrow_game():
  game_id = request.args.get('id')
  borrower_id = request.json.get('offeree_id')

  game = BoardGame.query.filter(BoardGame.id == game_id).first()

  # SET BORROW
  game.borrowed = True
  game.borrower_id = borrower_id

  #RESET LISTING
  game.forsale = False
  game.fortrade = False
  game.forborrow = False

  db.session.commit()
  return {"Message": "Game borrowed successfully."}

@games_routes.route('/returned', methods=["PATCH"])
def return_game():
  game_id = request.args.get('id')

  game = BoardGame.query.filter(BoardGame.id == game_id).first()

  game.borrowed = False
  game.borrower_id = None
  db.session.commit()
  return {"Message": "Game returned successfully."}

File: client/src/store/games.js
const GET_COLLECTION = 'games/get_collection';
const GET_GAMES_FOR_BUY = 'games/get_games_for_buy';
const GET_GAMES_FOR_TRADE = 'games/get_games_for_trade';
const GET_GAMES_FOR_BORROW = 'games/get_games_for_borrow';
const GET_GAMES_BY_TITLE = 'games/get_games_by_title';
const GET_GAME_FOR_OFFER = 'games/get_game_for_offer';
const GET_BORROWED_GAMES = 'games/get_borrowed_games';
const CLEAR_GAMES = 'games/clear_games';


export const getOfferGame = (game) => {
  return {
    type: GET_GAME_FOR_OFFER,
    game: game
  }
}

export const getUserCollection = (games) => {
  return {
    type: GET_COLLECTION,
    games: games
  };
};

export const getGamesForBuy = (games) => {
  return {
    type: GET_GAMES_FOR_BUY,
    games: games
  }
}

export const getGamesForTrade = (games) => {
  return {
    type: GET_GAMES_FOR_TRADE,
    games: games
  }
}

export const getGamesForBorrow = (games) => {
  return {
    type: GET_GAMES_FOR_BORROW,
    games: games
  }
}

export const getGamesByTitle = (games) => {
  return {
    type: GET_GAMES_BY_TITLE,
    games: games
  }
}

export const getBorrowed = (games) => {
  return {
    type: GET_BORROWED_GAMES,
    games: games
  }
}

export const clear = () => {
  return {
    type: CLEAR_GAMES,
  }
}

export const clearGamesState = () => {
  return async dispatch => {
    return dispatch(clear());
  }
}

export const getOffer = id => {
  return async dispatch => {
    const res = await fetch(`/api/games/offer?id=${id}`, {
      method: 'get'
    });
    res.data = await res.json();
    if (res.ok) {
      dispatch(getOfferGame(res.data.game));
    }
    return res;
  }
}

export const getCollection = id => {
  return async dispatch => {
    const res = await fetch(`/api/games/collection?id=${id}`, {
      method: 'get',
    });
    res.data = await res.json()
    if (res.ok) {
      dispatch(getUserCollection(res.data.games));
    }
    return res;
  }
}

export const getBorrowedGames = id => {
  return async dispatch => {
    const res = await fetch(`/api/games/borrowedgames?id=${id}`)
    res.data = await res.json()
    if (res.ok) {
      dispatch(getBorrowed(res.data.games));
    }
    return res;
  }
}

export const getForBuy = () => {
  return async dispatch => {
    const res = await fetch(`/api/games/forbuy`, {
      method: 'get',
    })
    res.data = await res.json();
    if (res.ok) {
      dispatch(getGamesForBuy(res.data.games))
    }
    return res
  }
}

export const getForTrade = () => {
  return async dispatch => {
    const res = await fetch(`/api/games/fortrade`, {
      method: 'get',
    })
    res.data = await res.json();
    if (res.ok) {
      dispatch(getGamesForTrade(res.data.games))
    }
    return res
  }
}

export const getForBorrow = () => {
  return async dispatch => {
    const res = await fetch(`/api/games/forborrow`, {
      method: 'get',
    })
    res.data = await res.json();
    if (res.ok) {
      dispatch(getGamesForBorrow(res.data.games))
    }
    return res
  }
}

export const getGamesByTitleToList = (id, searchTerm) => {
  return async dispatch => {
    const res = await fetch(`/api/games/bytitle?id=${id}&searchTerm=${searchTerm}`, {
      method: 'get',
    })
    res.data = await res.json();
    if (res.ok) {
      dispatch(getGamesByTitle(res.data.games))
    }
    return res;
  }
}

export const addGameToCollection = (user_id, game_id, username, title, year_published, thumb_url, msrp, listingPrice, rank, forsale, fortrade, forborrow, gameCondition, conditionDescription) => {
  return async dispatch => {
    try {
      const res = await fetch(`/api/games/add`, {
        method: 'post',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ user_id, game_id, username, title, year_published, thumb_url, msrp, listingPrice, rank, forsale, fortrade, forborrow, gameCondition, conditionDescription })
      })
      res.data = await res.json();
      if (res.ok) {
        dispatch(getCollection(user_id));
      }
      return res;
    } catch (err) {
      return { "Message": "Could not add game. Please check your submission and try again." }
    }
  }
}

export const addGameToSell = (user_id, game_id) => {
  return async dispatch => {
    try {
      const res = await fetch(`/api/games/toggleforsale`, {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ user_id, game_id })
      })
      res.data = await res.json();
      if (res.ok) {
        dispatch(getCollection(user_id))
      }
      return res;
    } catch (err) {
      return { "Message": "Could not patch game for sale." }
    }
  }
}

export const deleteGame = (user_id, game_id) => {
  return async dispatch => {
    try {
      const res = await fetch(`/api/games/remove?game_id=${game_id}`, {
        method: 'DELETE',
      })
      res.data = await res.json();
      if (res.ok) {
        dispatch(getCollection(user_id))
      }
      return res;
    } catch (err) {
      return { "Message": "Count not remove game." }
    }
  }
}

export const updateGame = (user_id, game_id, listingPrice, gameCondition, conditionDescription, forsale, fortrade, forborrow) => {
  return async dispatch => {
    try {
      const res = await fetch(`/api/games/update?game_id=${game_id}`, {
        method: "PATCH",
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ game_id, listingPrice, gameCondition, conditionDescription, forsale, fortrade, forborrow })
      })
      res.data = await res.json();
      if (res.ok) {
        dispatch(getCollection(user_id));
      }
      return res;
    } catch (err) {
      return { "Message": "Could not update game." }
    }
  }
}

export const changeOwner = (user_id, offeree_id, username, game_id) => {
  return async dispatch => {
    try {
      const res = await fetch(`/api/games/changeowner`, {
        method: "PATCH",
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ offeree_id, username, game_id })
      });
      res.data = await res.json();
      if (res.ok) {
        dispatch(getCollection(user_id))
      }
      return res;
    } catch (err) {
      return { "Message": "Could not complete transaction." }
    }
  }
}

export const borrowed = (user_id, offeree_id, game_id) => {
  return async dispatch => {
    const res = await fetch(`/api/games/borrow?id=${game_id}`, {
      method: "PATCH",
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ offeree_id })
    });
    res.data = await res.json();
    if (res.ok) {
      dispatch(getCollection(user_id));
    }
    return res;
  }
}

export const returned = (user_id, game_id) => {
  return async dispatch => {
    const res = await fetch(`/api/games/returned?id=${game_id}`, {
      method: "PATCH",
      headers: {
        'Content-Type': 'application/json'
      },
    });
    res.data = await res.json()
    if (res.ok) {
      dispatch(getCollection(user_id))
    }
  }
}


// export const getCollection = id => {
//   return async dispatch => {
//     const res = await fetch(`/api/games/collection?id=${id}`, {
//       method: 'get',
//     })
//     res.data = await res.json()
//     if (res.ok) {
//       dispatch(getUserCollection(res.data.games));
//     }
//     return res;
//   }
// }

export default function gamesReducer(state = {}, action) {
  switch (action.type) {
    case GET_COLLECTION:
      return { ...state, collection: action.games };
    case GET_BORROWED_GAMES:
      return { ...state, borrowedGames: action.games };
    case GET_GAME_FOR_OFFER:
      return { ...state, game: action.game };
    case GET_GAMES_FOR_BUY:
      return { ...state, gamesforbuy: action.games };
    case GET_GAMES_FOR_TRADE:
      return { ...state, gamesfortrade: action.games };
    case GET_GAMES_FOR_BORROW:
      return { ...state, gamesforborrow: action.games };
    case GET_GAMES_BY_TITLE:
      return { ...state, gamesByTitle: action.games };
    case CLEAR_GAMES:
      return {}
    default:
      return state;
  }
}
